06x-new-kafka-roller.md

```diff
@@ -0,0 +1,331 @@
+
+      - At this point either we started nodes or decided not to because nodes did not have `POD_HAS_OLD_REVISION` reason. Regardless, wait for nodes to have `SERVING` within `postOperationalTimeoutMs`. If the timeout is reached and the node's `numRetries` is greater than or equal to `maxRetries`, throw `TimeoutException`. Otherwise increment node's `numRetries` and repeat from step 2. 
+
+   
+   - Otherwise the nodes will be attempted to restart one by one in the following order:
+      - Pure controller nodes
+      - Combined nodes
+      - Broker only nodes
+      
+   - Wait for the restarted node to have `SERVING` within `postOperationalTimeoutMs`. If the timeout is reached and the node's `numRetries` is greater than or equal to `maxRetries`, throw `TimeoutException`. Otherwise increment node's `numRetries` and repeat from step 2. 
+
+7. Further refine the broker nodes in `MAYBE_RECONFIGURE` group:
+   - Describe Kafka configurations for each node via Admin API and compare them against the desired configurations. This is essentially the same mechanism we use today for the current KafkaRoller.
+   - If a node has configuration changes and they can be dynamically updated, add the node into another group called `RECONFIGURE`.
+   - If a node has configuration changes but they cannot be dynamically updated, add the node into the `RESTART` group.
+   - If a node has no configuration changes, put the node into the `NOP` group.
+
+8. Reconfigure each node in `RECONFIGURE` group:
+   - If `numReconfigAttempts` of a node is greater than the configured `maxReconfigAttempts`, add a restart reason to its context and repeat from step 2. Otherwise continue.
+   - Send `incrementalAlterConfig` request with its config updates.
+   - Transitions the node's state to `RECONFIGURED` and increment its `numReconfigAttempts`.
+   - Wait for each node that got configurations updated until they have `LEADING_ALL_PREFERRED` within the `postOperationalTimeoutMs`.
+   - If the `postOperationalTimeoutMs` is reached, repeat from step 2.
+
+9. If at this point, the `RESTART` group is empty and if there is no nodes that is in `NOT_READY` state, the reconciliation will be completed successfully.
+   - If there are nodes in `NOT_READY` state, wait for them to have `SERVING` within the `postOperationalTimeoutMs`. 
+   - If the timeout is reached for a node and its `numRetries` is greater than or equal to `maxRetries`, throw `TimeoutException`. 
+   - Otherwise increment node's `numRetries` and repeat from step 2.
+   This is consistent with how the current roller handles unready nodes.
+
+10. Otherwise, batch nodes in `RESTART` group and get the next batch to restart:
+   - Further categorize nodes based on their roles so that the following restart order can be enforced:
+       1. `NON_ACTIVE_CONTROLLER` - Pure controller that is not the active controller
+       2. `ACTIVE_CONTROLLER` - Pure controller that is the active controller (the quorum leader)
```

Comment:
Why not always roll the active controller last, whether it is in combined mode or not?